python/cuspatial/cuspatial/core/binops/binops.py

```diff
@@ -0,0 +1,409 @@
+        - polygon point
+        - polygon multipoint
+        - polygon linestring
+        - polygon multilinestring
+        - polygon polygon
+        - polygon multipolygon
+        - multipolygon point
+        - multipolygon multipoint
+        - multipolygon linestring
+        - multipolygon multilinestring
+        - multipolygon polygon
+        - multipolygon multipolygon
+
+        Parameters
+        ----------
+        op : str
+            The binary operation to perform.
+        lhs : GeoSeries
+            The left-hand-side of the binary operation.
+        rhs : GeoSeries
+            The right-hand-side of the binary operation.
+        align : bool
+            If True, align the indices of `lhs` and `rhs` before performing
+            the binary operation. If False, `lhs` and `rhs` must have the
+            same index.
+
+        Returns
+        -------
+        GeoSeries
+            A GeoSeries containing the result of the binary operation.
+        """
+        (self.lhs, self.rhs) = lhs.align(rhs) if align else (lhs, rhs)
+        self.align = align
+
+        # Type disambiguation
+        # Type disambiguation has a large effect on the decisions of the
+        # algorithm.
+        (lhs, rhs, indices) = self.preprocess(op, self.lhs, self.rhs)
+
+        # Binop call
+        _binop = getattr(self, op)
+        point_result = _binop(lhs, rhs)
+
+        # Postprocess: Apply discrete math rules to identify relationships.
+        final_result = self.postprocess(op, indices, point_result)
+
+        # Return
+        self.op_result = final_result
+
+    def __call__(self) -> cudf.Series:
+        """Return the result of the binary operation."""
+        return self.op_result
+
+    def contains_properly(self, lhs, points):
+        """Compute the contains_properly relationship between two GeoSeries.
+        A feature A contains another feature B if no points of B lie in the
+        exterior of A, and at least one point of the interior of B lies in the
+        interior of A. This is the inverse of `within`."""
+        if contains_only_points(lhs) and contains_only_points(points):
+            return self.equals(lhs, points)
+        if not contains_only_polygons(lhs):
+            raise TypeError(
+                "`.contains` can only be called with polygon series."
+            )
+
+        # call pip on the three subtypes on the right:
+        point_result = contains_properly(
+            points.x,
+            points.y,
+            lhs.polygons.part_offset[:-1],
+            lhs.polygons.ring_offset[:-1],
+            lhs.polygons.x,
+            lhs.polygons.y,
+        )
+        return point_result
+
+    def equals(self, lhs, rhs):
+        """Compute the equals relationship between two GeoSeries."""
+        if contains_only_points(lhs):
+            result = lhs.points.xy.equals(rhs.points.xy)
+        elif contains_only_linestrings(lhs):
+            result = lhs.lines.xy.equals(rhs.lines.xy)
```

Comment:
Also, the points in both geometries may not be in the same order...

Reply:
I am writing a test for this now, what orderings of LineStrings GeoPandas considers to be equal.

Reply:
Hmm. Now that I dig into this deeper I'm reminded that `equals` is in the next PR, not this one. I think I got ahead of myself on the PR and wrote functionality that isn't intended to be tested in this branch.

Reply:
Equals functionality is in https://github.com/rapidsai/cuspatial/issues/838. Do you want  me to carefully remove the equals functionality from this branch, or do you mind ignoring it until the next PR?

Reply:
Removed.